site/e2e/tests/organizations/customRoles/customRoles.spec.ts

```diff
@@ -0,0 +1,201 @@
+		const menu = page.locator("#more-options");
+		await menu.getByText("Edit").click();
+
+		await expect(page).toHaveURL(
+			`/organizations/${org.name}/roles/${customRole.name}`,
+		);
+
+		const cancelButton = page.getByRole("button", { name: "Cancel" }).first();
+		await expect(cancelButton).toBeVisible();
+		await cancelButton.click();
+
+		await expect(page).toHaveURL(`/organizations/${org.name}/roles`);
+
+		await deleteOrganization("developers");
+	});
+
+	test("create custom role, edit role and save changes", async ({ page }) => {
+		requiresLicense();
+		await setupApiCalls(page);
+
+		const org = await createOrganizationWithName("users");
+
+		const customRole = await createCustomRole(
+			org.id,
+			"custom-role-test-1",
+			"Custom Role Test 1",
+		);
+
+		await page.goto(`/organizations/${org.name}/roles`);
+		const roleRow = page.getByTestId(`role-${customRole.name}`);
+		await expect(roleRow.getByText(customRole.display_name)).toBeVisible();
+		await expect(roleRow.getByText("organization_member")).toBeVisible();
+
+		await page.goto(`/organizations/${org.name}/roles/${customRole.name}`);
+
+		const displayNameInput = page.getByRole("textbox", {
+			name: "Display name",
+		});
+		await displayNameInput.fill("Custom Role Test 2 Edited");
+
+		const groupCheckbox = page.getByTestId("group").getByRole("checkbox");
+		await expect(groupCheckbox).toBeVisible();
+		await groupCheckbox.click();
+
+		const organizationMemberCheckbox = page
+			.getByTestId("organization_member")
+			.getByRole("checkbox");
+		await expect(organizationMemberCheckbox).toBeVisible();
+		await organizationMemberCheckbox.click();
+
+		const saveButton = page.getByRole("button", { name: "Save" }).first();
+		await expect(saveButton).toBeVisible();
+		await saveButton.click();
+
+		await expect(roleRow.getByText("Custom Role Test 2 Edited")).toBeVisible();
+
+		const roleRow2 = page.getByTestId(`role-${customRole.name}`);
+		await expect(roleRow2.getByText("organization_member")).not.toBeVisible();
+		await expect(roleRow2.getByText("group")).toBeVisible();
+
+		await expect(page).toHaveURL(`/organizations/${org.name}/roles`);
+
+		await deleteOrganization("users");
+	});
+
+	test("displays built-in role without edit/delete options", async ({
+		page,
+	}) => {
+		requiresLicense();
+		await setupApiCalls(page);
+
+		const org = await createOrganizationWithName("testers");
+
+		await page.goto(`/organizations/${org.name}/roles`);
+
+		const roleRow = page.getByTestId("role-organization-admin");
+		await expect(roleRow).toBeVisible();
+
+		await expect(roleRow.getByText("Organization Admin")).toBeVisible();
+
+		// Verify that the more menu (three dots) is not present for built-in roles
+		await expect(
+			roleRow.getByRole("button", { name: "More options" }),
+		).not.toBeVisible();
+
+		await deleteOrganization("testers");
+	});
+
+	test("create custom role with UI", async ({ page }) => {
```

Comment:
why not make this a helper and call it from the other test rather than using the api?

Reply:
I have actually been going the route of having 2 different types of tests, one that starts with a change using the api and then completing the test through the UI and then another that does the same thing through the UI. I think there is the potential catching edge cases through this way of testing.

Reply:
@aslilac Looking around at many of the existing tests, I see this pattern when the same things to happen for many tests, for example creating a template, the API is used but then atleast one test would do the same thing through the UI if possible. I feel like this many be a bit more performant than going through the UI for every test. The except would be if there are many variations of path the user could take through the UI.